exo/inference/torch/tests/test_split_model.py

```diff
@@ -0,0 +1,378 @@
+                cache_position,
+                past_key_values,
+                output_attentions
+            )
+
+            #print(f"causal_mask.dim(): {self.causal_mask.dim()}")
+
+            print(f"\ncausal_mask:{self.causal_mask}\n\n")
+
+            # embed positions, some models require and some dont
+            if isinstance(model, LlamaModel):
+                self.position_embeddings = model.rotary_emb(
+                    self.inputs_embeds,
+                    position_ids
+                )
+ 
+            model_inputs = llm_model.prepare_inputs_for_generation(
+                self.input_ids,
+                past_key_values=past_key_values,
+                attention_mask=self.attention_mask,
+                inputs_embeds=self.inputs_embeds,
+                position_ids=position_ids,
+                cache_position=cache_position
+            )
+
+            print(f"model_inputs\n{model_inputs}")
+
+            self.hidden_states = self.inputs_embeds
+            self.position_ids = model_inputs["position_ids"]
+            self.cache_position = model_inputs["cache_position"]
+            self.past_key_values = model_inputs["past_key_values"]
+
+
+        for decoder_layer in self.layers:
+            layer_outputs = decoder_layer(
+                self.hidden_states,
+                attention_mask=self.causal_mask,
+                position_ids=self.position_ids,
+                past_key_values=self.past_key_values,
+                use_cache=True,
+                cache_position=self.cache_position
+
+            )
+
+            self.hidden_states = layer_outputs[0]
+            self.next_decoder_cache = layer_outputs[1]
+
+        if self.is_last:
+            self.hidden_states = model.norm(self.hidden_states)
+
+            if use_legacy_cache:
+                self.past_key_values = self.next_decoder_cache.to_legacy_cache()
+            else:
+                self.past_key_values = self.next_decoder_cache
+            
+            # lm_head  
+            logits = llm_model.lm_head(self.hidden_states).to("cuda")
+
+            return (
+                None,
+                None,
+                logits
+            )
+        
+        return (
+            self.hidden_states,
+            self.past_key_values,
+            None
+        )
+
+async def model_half_split_test(prompt: str, model_id: str, layers: int):
+    """
+    Test for splitting in half
+    """
+
+    half_layers = int(layers / 2)
+
+    # inference
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    max_length = 512 #tokenizer.model_max_length
+
+    # get llm model
+    llm_model = AutoModelForCausalLM.from_pretrained(
+        model_id,
+        torch_dtype="auto",
+        device_map="auto",
+        use_cache=True
+    )
+    
+    # get base model
+    model = llm_model.model
+
+    # add pad token if none, depending on model
+    if tokenizer.pad_token == None:
+        if re.match(r"Llama|llama", model_id):
```

Comment:
This was because of an issue of the padding token missing in the llama3 model [1](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/discussions/101) [2](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/discussions/40) which I added to see if I could fix some generation issues. I don't think its needed and will test without. This is more for training.